### ISSUE TITLE
Feature request: moving datasets

### ISSUE THREAD

**Issue by clarfonthey** [openzfs/zfs#13393]
### Describe the feature would like to see added to OpenZFS

I would like to see the ability to live move datasets. As a placeholder syntax, something like `zfs move path/to/old path/to/new` would move the dataset from `path/to/old` to `path/to/new` without taking the dataset offline. The options of the new dataset would match the old one, overriding defaults as necessary.

As a useful first step, this could be enabled only for offline (unmounted) datasets. Both the old and new datasets would not be able to be mounted until the data is copied from the old one to the new one, and the old one is destroyed.

Presumably, this would work recursively, transferring snapshots and child datasets.

As an even smaller first step, this could be done for datasets who don't modify the hierarchy of datasets, e.g. from `pool/parent/child1` to `pool/parent/child2`. Since these moves don't require actually copying any data, the dataset could simply be renamed while preserving the mount point.

### How will this feature improve OpenZFS?

Right now, there is no real way to perform an `mv` operation with datasets. In cases where the hierarchy of datasets isn't changed (e.g. `pool/old` to `pool/new`) this could be done safely by simply renaming the dataset, but if the hierarchy is changed, the dataset has to be copied over.

From a user perspective, the only way to achieve this is to create a new dataset, copy the data over, and destroy the old dataset. This can be difficult when moving large datasets, as you have to take care to make sure the amount of data duplicated between both datasets does not exceed the capacity of the drive. If this had support within zfs itself, we could mitigate these concerns automatically.

Additionally, it would be extremely desirable to be able to rename or move datasets that are currently in use, for example, moving `/var/log` from one physical drive to another. Since it's difficult to configure an existing system to unmount these drives without booting into another system (it can be done, since it's done at shutdown, but it's a huge pain), letting the user simply move these datasets over would be incredibly useful. Obviously, this substantially increases the complexity of the implementation, and the ability to do so for offline datasets would be a useful first step.

### Additional context

This would likely be implemented via the following flow, as a basic sketch:

1. Set a quota on the old dataset to ensure that it cannot become larger than the space of the target dataset (e.g. moving from a large volume to a smaller one). This could potentially be configurable, with the quota being the size of the data plus some "wiggle room" for extra writes.
2. Pause all dataset-related operations on the source dataset.
3. Create the new dataset with the same options as the old one, also with an inability to perform dataset-related operations. Make sure the dataset is unmounted and reserved with an adequate amount of space.
4. Start copying over data from the old dataset to the new dataset.
5. During this process, writes to the dataset are still allowed, but they may need to be throttled to ensure that the copying of old data outpaces the writing of new data.
6. Once the datasets are synced, pause all I/O operations on the old dataset.
7. If the dataset is mounted, modify all open files to point to the new dataset.
8. Unmount the old dataset and put the new dataset in its place.
9. Unpause I/O.
10. Destroy the old dataset.

I think I've seen discussions of this before, but did a quick search and saw no open features here on it. Apologies if there is some existing discussion somewhere else that I didn't find.

**Comment by nabijaczleweli** [comment]
How's this different from `zfs rename` (or, for the simpler case, `zfs set mountpoint=`)?

**Comment by rincebrain** [comment]
Or `mount --move`.

I suspect the functionality desired here could be implemented with purely userland work, since it seems like they'd like a readonly version at the old mountpoint and a readwrite at the end, and since we allowed multiple mounting of a dataset, we could probably do it...

**Comment by clarfonthey** [comment]
So, when I created this, I for some reason had no idea that `zfs rename` existed. I guess that I had been relying on knowledge since before it did, and then just for some reason never found out when it was implemented.

At some point I'll reword the OP to make this issue specifically about moving online datasets, rather than offline ones, since that's the thing that doesn't exist right now.

I was under the impression that something along these lines existed in some other filesystem, but I could have been mistaken. I guess that everything right now is flexible in terms of the actual data moving on disk (for example, via replication or devices being added/removed from a pool) but not the data moving between physical mounts. I'm not actually sure how file handlers are done in ZFS on Linux, and whether they're forced to be linked to the specific dataset or can actually be moved within a zpool. Because most of my use cases are things like moving `pool/log` to `pool/data/log`, etc., with the dataset being mounted at `/var/log`.

**Comment by clarfonthey** [comment]
Yeah, I must have just been compounding misunderstandings when I didn't know that `zfs rename` existed, or that it allowed changing the nesting of datasets.

Anyway, updated the original description to reference `zfs rename` and `mount --move`, so that hopefully it's a bit clearer what's being requested.

**Comment by GregorKopka** [comment]
I would be happy if `zfs rename` would not remount a filesystem in case the effective mountpoint does not change.

**Comment by freqlabs** [comment]
Is `zfs rename -u` what you're looking for?

**Comment by GregorKopka** [comment]
> Is `zfs rename -u` what you're looking for? I suppose you might want it automatically...

Apart from having missed that this option exists by now (thanks for the info):
Yes.

An operation that is a effective NOP should IMHO not have side effects.

**Comment by clarfonthey** [comment]
Oh huh, glad to know that `zfs rename -u` exists now.

This is a bit of an awkward side effect though, since can't ZFS perform some kind of `mount --move` operation when setting `mountpoint` normally? It would be nice to combine this effect in the command somehow, although I'm not sure if that just transparently unmounts and remounts instead.

**Comment by rincebrain** [comment]
It unmounts/remounts, AFAIK.

**Comment by clarfonthey** [comment]
I don't have the time to look into this right now, but if that's the case it may make more sense to close this issue and open an issue specifically on using `mount --move` instead of unmounting & remounting volumes when changing mount points. If someone else wants to open an issue elaborating on that, feel free, otherwise I'll do it when I get around to it.

**Comment by rincebrain** [comment]
Seems reasonable to me.
